### PR TITLE
Wrap callable resolution

### PR DESCRIPTION
**Pull request by lalop**
Wrap the callable resolution to resolve it when it's called

### CODE REVIEW

**Comment by JoeBengalen** [comment]
You are mixing `to_resolve` and `toResolve`

**Comment by lalop** [comment]
Oh yes, 
thank you

**Comment by lalop** [comment]
I will be happy to have some review on this 
\cc @codeguy @akrabat

**Comment by codeguy** [comment]
Deferring to @akrabat on this.

**Comment by akrabat** [comment]
What does the change give us and how does it affect what's happening in #822 ? 

Also you'll need to rebase and take into account the change in #1184

**Comment by akrabat** [comment]
OK, I've read #1178 now, so know why this PR exists. Just need the rebase with the #1184 change taken into account.

**Comment by lalop** [comment]
Fun fact, I have added a test that pass thanks to #1178 but as a side effect of this patch.
We never checked if the string past to the resolver is callable.
Can we list all the callable things we have : 
- `closure`
- `'function name'`
- `[$obj, $method]`
- `'Class:method'`
- `'container_key:method'`

Maybe we should add `container_key` that can be a function after all.

Other point, I'm not so happy with the name I gave to the class name `CallableResolver` in front of `ResolveCallable` for the trait, any inspiration is welcome.

File: Slim/CallableResolver.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    
+    protected $resolved;
+    
+    public function __construct($toResolve, Container $container)
+    {
+        $this->toResolve = $toResolve;
+        $this->container = $container;
+    }
+    
+    public function resolve()
```
Comment (mnapoli) [inline Slim/CallableResolver.php:22]: @lalop why is this method public? I don't see it being called anywhere except in this class.

Comment (lalop) [inline Slim/CallableResolver.php:22]: yes, good point.
I've fixed that in an other pr https://github.com/slimphp/Slim/pull/1188
Thank you

Comment (mnapoli) [inline Slim/CallableResolver.php:22]: :+1:

**Pull request merged**